### PR TITLE
nzedb pre import

### PR DESCRIPTION
**Pull request by brookesy2**
I only made this PR to chat about the below code.

The issue I have is that pre.name is a unique column, if I use COPY in postgres and a single line fails, the whole CSV fails, not to mention the fact that a release may get nuked later, so I really need to perform an upsert.

Any thoughts on ways to do this? I was thinking of doing line by line and only committing after 100 or so lines, still have the same issue though, the need to upsert.

Edit: Another thought, I could do it similar to how pre.py works, but I would need to requery the pre table after each iteration (or however many before I commit)

### CODE REVIEW

**Comment by Murodese** [comment]
COPY is a fast insert method for a good reason: it does a direct memory map between the file you give and the table schema, and ignores all keys. Hence, you can only really use it if you're very carefully sanitising the input. Pynab can use it for segments/parts because we very carefully check if there are duplicates existing prior to running the copy. We do this by pulling all already-collected data and merging it with the new data we're inserting, then only COPY the stuff that's missing. Because it's only done in chunks of 500 parts (~~25k segments or so at most), there's no memory problems.

Basically, pull out any relevant data in batches first and check for duplicates in Python, because you can't do it during a COPY. It's also still a lot faster to do that than to insert line-by-line.

**Comment by brookesy2** [comment]
You make it sound so simple :) now just yo code it! At least I learn a lot!!

Another oddity, see that last sed command? Works fine in shell, does not
work when executed in python. Very odd.

Annoyingly the data has \N as the last  column... (As well as new line)

**Comment by brookesy2** [comment]
So I was all excited because I got things working with pandas, clean CSVs, happy days. Turns out I forgot the fact that a later CSV may contain a nuke for a previous CSV... :(

Back to the drawing board!

**Comment by brookesy2** [comment]
Figured it out, I will just delete any matching rows from the pre table prior to inserting the new CSV. I guess the only bad thing that could happen is if the pre is already linked to a release.

**Comment by brookesy2** [comment]
@Murodese You able to give this a try? It imported around 260k pre's no problem. Was fairly fast too!

**Comment by Murodese** [comment]
Sorry, been busy:

```
james@indexer:/opt/pynab$ python3 scripts/nzedb_pre_import.py
  File "scripts/nzedb_pre_import.py", line 117
    print("Pre-Import: Attempting to add {} to the database".format(processingFile['lastfile'])))
                                                                                                ^
SyntaxError: invalid syntax
```

**Comment by Murodese** [comment]
Also, add pandas/bs4 to requirements.txt if you want to use them.

Also, missing:

`sys.path.append(os.path.join(os.path.dirname(os.path.realpath(__file__)), '..'))`

Above the pynab imports, otherwise `python3 scripts/<script>.py` doesn't work :)

**Comment by brookesy2** [comment]
Eh, such is life :)

Sorry about that, please give it a try now!

**Comment by Murodese** [comment]
`sh: 1: csvcut: not found`

**Comment by Murodese** [comment]
Need to add csvkit to requirements.txt as well.

**Comment by Murodese** [comment]
Lots of errors:

```
Pre-Import: Attempting to download file: 1404840222
Pre-Import: Attempting to add 1404840222 to the database
Pre-Import: Error inserting into database - duplicate key value violates unique constraint "pres_pkey"
DETAIL:  Key (id)=(11271) already exists.
CONTEXT:  COPY pres, line 1: "Star.Trek.TNG.S06E05.In.den.Subraum.entfuehrt.German.DL.FS.1080p.BluRay.x264-RSG,,0,X264,2014-07-08 ..."

Pre-Import: Attempting to download file: 1404926624
Pre-Import: Attempting to add 1404926624 to the database
Pre-Import: Error inserting into database - duplicate key value violates unique constraint "pres_pkey"
DETAIL:  Key (id)=(11272) already exists.
CONTEXT:  COPY pres, line 1: "Die.schoensten.Kuesten.Frankreichs.E04.Von.Douarnenez.nach.Saint.Nazaire.German.DOKU.WS.HDTVRip.x264..."

Pre-Import: Attempting to download file: 1405013027
Pre-Import: Attempting to add 1405013027 to the database
Pre-Import: Error inserting into database - duplicate key value violates unique constraint "pres_pkey"
DETAIL:  Key (id)=(11273) already exists.
CONTEXT:  COPY pres, line 1: "Mohikana-I_Will_Find_You-(861_527-2)-CDM-FLAC-1996-B2A,,0,FLAC,2014-07-10 13:08:00,#a.b.flac,46868,a..."

Pre-Import: Attempting to download file: 1405099430
Pre-Import: Attempting to add 1405099430 to the database
Pre-Import: Error inserting into database - duplicate key value violates unique constraint "pres_pkey"
DETAIL:  Key (id)=(11274) already exists.
CONTEXT:  COPY pres, line 1: "Esteban-Just_Doin_Me-(Promo_CDS)-2014-YVP_INT,,0,MP3,2014-07-11 03:39:35,orlydb,365842,alt.binaries...."
```

**Comment by brookesy2** [comment]
Oh man. I swear this was pain free!! I will investigate today. Thanks for
ya patience!!

**Comment by brookesy2** [comment]
I'm more surprised you managed to find 4 things that weren't porn.

**Comment by Murodese** [comment]
The fifth was porn, so I scaled it back a bit :p

**Comment by brookesy2** [comment]
Ha!

Trying to rewrite without csv kit, might take a bit. Saves needing more things in requirements.

**Comment by Murodese** [comment]
Not a big deal, no need to reinvent the wheel.

**Comment by brookesy2** [comment]
Its all good, pandas does most of the same things anyway, so makes sense. This is the most hilariously over engineered import process ever :)

It's working, just running it a few times to make sure it keeps going. Will check in shortly!

**Comment by brookesy2** [comment]
Ok give that a try, I did a full import, which was fine, also did a part import with existing pres. Hopefully no errors for you!

Also in lines 101/102 is there a fast way to delete the result of a query?

**Comment by Murodese** [comment]
`db.query(Pres).filter(Pres.id.in_(pres)).delete(synchronize_session='fetch')`

**Comment by Murodese** [comment]
```
Pre-Import: Attempting to download file: 1406395470
Pre-Import: Attempting to add 1406395470 to the database
Pre-Import: Error inserting into database - duplicate key value violates unique constraint "pres_pkey"
DETAIL:  Key (id)=(11289) already exists.
CONTEXT:  COPY pres, line 1: "MilfNextDoor.14.07.26.Brianna.Ray.And.Krystal.Carrington.Make.Out.Moms.XXX.MP4-FUNKY,MND260714XAAGH3..."

Pre-Import: Attempting to download file: 1406481873
Pre-Import: Attempting to add 1406481873 to the database
Pre-Import: Error inserting into database - duplicate key value violates unique constraint "pres_pkey"
DETAIL:  Key (id)=(11290) already exists.
CONTEXT:  COPY pres, line 1: "Dorothy_L._Sayers_-_Der_Mann_Ohne_Gesicht-Audiobook-DE-2003-oNePiEcE_INT,,0,ABOOK,2014-07-27 10:28:4..."

Pre-Import: Attempting to download file: 1406568275
Pre-Import: Error downloading: https://www.dropbox.com/sh/fb2pffwwriruyco/AACN4HaTBmfYNJJG_gQjkn9Ia/1406568275_predb_dump.csv.gz?dl=1
Failures: ['1406395470', '1406481873', '1406568275']
```

**Comment by Murodese** [comment]
For some context, my predb just has some stuff in it from that sql you linked me a couple of days ago, plus the IRC bot.

**Comment by brookesy2** [comment]
Thanks for the context. Time to investigate why I am not getting that error. It looks like its not correctly removing the already existing pres.

**Comment by brookesy2** [comment]
Hrmm, so weird. Its cruising along fine on both my prod db and test db.

In line 110 can you put print(pre). I just want to see if its actually deleting them.

**Comment by Murodese** [comment]
```
Pre-Import: Attempting to download file: 1406568275
<pynab.db.Pre object at 0x7f4a4636ef98>
Pre-Import: Attempting to add 1406568275 to the database
Pre-Import: Error inserting into database - duplicate key value violates unique                 constraint "pres_pkey"
DETAIL:  Key (id)=(11291) already exists.
CONTEXT:  COPY pres, line 1: "CrazyBabe.09.03.xx.Scar.Take.Care.Rubber.Ducky.XXX                .HR.WMV-OHRLY,ohrly-cb903stcrd,0,XXX,2014-07-27 17..."

Pre-Import: Attempting to download file: 1406654678
Pre-Import: Attempting to add 1406654678 to the database
Pre-Import: Error inserting into database - duplicate key value violates unique                 constraint "pres_pkey"
DETAIL:  Key (id)=(11292) already exists.
CONTEXT:  COPY pres, line 1: "Seth.Meyers.2014.07.24.Hank.Azaria.720p.HDTV.x264-                aA,,0,TV-X264,2014-07-29 12:52:42,prelist,522338,a..."

Pre-Import: Attempting to download file: 1406741080
Pre-Import: Attempting to add 1406741080 to the database
Pre-Import: Error inserting into database - duplicate key value violates unique                 constraint "pres_pkey"
DETAIL:  Key (id)=(11293) already exists.
CONTEXT:  COPY pres, line 1: "Dragon.Ball.Z.S03E26.1080p.WS.BluRay.x264-CCAT,,1,                X264,2014-07-29 16:26:11,#pre@corrupt,0,"

Pre-Import: Attempting to download file: 1406827483
^CPre-Import: Error downloading: https://www.dropbox.com/sh/fb2pffwwriruyco/AABA                PuablfVZHt10I40LAnzga/1406827483_predb_dump.csv.gz?dl=1
Failures: ['1406568275', '1406654678', '1406741080', '1406827483']
```

**Comment by brookesy2** [comment]
Lets try print(pre.name) so we can see if its the same one as the error. First file it looks like it did find a duplicate, second one, it didnt.

Do you use IRC? Should make a pynab channel on freenode, might join one now!

**Comment by Murodese** [comment]
```Pre-Import: Attempting to download file: 1406827483
High.Heels.And.Glasses.2.XXX.720p.WEBRiP.WMV-TBP
High.Heels.And.Glasses.3.XXX.720p.WEBRiP.WMV-TBP
Pre-Import: Attempting to add 1406827483 to the database
Pre-Import: Error inserting into database - duplicate key value violates unique constraint "pres_pkey"
DETAIL:  Key (id)=(11294) already exists.
CONTEXT:  COPY pres, line 1: "Paris.A.Tout.Prix.2013.FRENCH.COMPLETE.BLURAY-MODERN,,0,BLURAY,2014-07-31 13:07:56,#pre@corrupt,0,"

Pre-Import: Attempting to download file: 1406913886
Pre-Import: Attempting to add 1406913886 to the database
Pre-Import: Error inserting into database - duplicate key value violates unique constraint "pres_pkey"
DETAIL:  Key (id)=(11295) already exists.
CONTEXT:  COPY pres, line 1: "Gefangene.Frauen.1980.German.DL.1080p.BluRay.x264-iFPD,,0,X264,2014-08-01 12:29:12,prelist,0,"

Pre-Import: Attempting to download file: 1407000288
Pre-Import: Attempting to add 1407000288 to the database
Pre-Import: Error inserting into database - duplicate key value violates unique constraint "pres_pkey"
DETAIL:  Key (id)=(11296) already exists.
CONTEXT:  COPY pres, line 1: "Czas.Honoru.Powstanie.S01E01.PL.HDTV.x264-A4O,,0,TV-XVID,2014-08-02 13:24:07,#pre@corrupt,0,"
```

Doesn't appear to be the same one.

I'm on IRC most of the time, synirc though. If there's some interest I'm happy to hang out there, it's just easier to get back to stuff here regardless of timezones though.

**Comment by brookesy2** [comment]
That is super weird. Can you dump out your pre table so I can use it? Hopefully I can repro it locally so you don't have to wait around so long!

True, does make it easier on delayed responses here.

**Comment by brookesy2** [comment]
Found a bug in prebot I need to fix because of this!

**Comment by Murodese** [comment]
https://www.dropbox.com/s/hw8xrsmfuow2jum/pres.zip?dl=0

**Comment by brookesy2** [comment]
Weirdly that file gives me a ton of syntax errors. Will figure it out.

**Comment by brookesy2** [comment]
I think I see why. A bunch of \N's are in there. Which from original imports is what it used as a null value.

If you filter those from your pre table can you check if the import works?

**Comment by Murodese** [comment]
Oh, probably because it was a copy backup. Try this: https://www.dropbox.com/s/hw8xrsmfuow2jum/pres.zip?dl=0 (no newlines in that)

**Comment by brookesy2** [comment]
Worked like a charm. Time to go to town on finding this issue!

**Comment by brookesy2** [comment]
Ok, this is odd, the in_ query for whatever reason doesn't work on your pre table, but works on mine... I checked and they look identical. Below is my table that works. Notice anything odd? Looks the same as yours to me!

```sql
-- Table: pres

-- DROP TABLE pres;

CREATE TABLE pres
(
  id bigserial NOT NULL,
  pretime timestamp without time zone,
  name character varying,
  searchname character varying,
  category character varying,
  source character varying,
  requestid bigint,
  requestgroup character varying,
  filename character varying,
  nuked boolean,
  CONSTRAINT pres_pkey PRIMARY KEY (id),
  CONSTRAINT pres_name_key UNIQUE (name)
)
WITH (
  OIDS=FALSE
);
ALTER TABLE pres
  OWNER TO pynab;

-- Index: ix_pres_name

-- DROP INDEX ix_pres_name;

CREATE INDEX ix_pres_name
  ON pres
  USING btree
  (name COLLATE pg_catalog."default");

-- Index: ix_pres_requestgroup

-- DROP INDEX ix_pres_requestgroup;

CREATE INDEX ix_pres_requestgroup
  ON pres
  USING btree
  (requestgroup COLLATE pg_catalog."default");

-- Index: ix_pres_requestid

-- DROP INDEX ix_pres_requestid;

CREATE INDEX ix_pres_requestid
  ON pres
  USING btree
  (requestid);

```

**Comment by Murodese** [comment]
My ID field is bigint, not bigserial.

**Comment by Murodese** [comment]
Did you create those tables by hand and create the model afterwards? Sqlalchemy doesn't support the use of bigserial.

**Comment by brookesy2** [comment]
Very odd. I pretty sure I used alembic, I guess not. Well, now to figure out why the in_ returns 0 results!

**Comment by brookesy2** [comment]
Err, I am reading this error wrong? I checked the file it tries to COPY and it does not contain the name it complains about. Does the below seem to indicate that its somehow trying to create the same primary key again?

Pre-Import: Error inserting into database - duplicate key value violates unique constraint "pres_pkey"
 DETAIL:  Key (id)=(15347) already exists.
 CONTEXT:  COPY pres, line 1

**Comment by brookesy2** [comment]
So if I truncate the pres table you sent, and try the import, its fine. So maybe some anomaly in your pres table sequence?

**Comment by Murodese** [comment]
Do the dropbox files you grab contain inserts with primary key IDs already in the file? That would cause problems in itself. The PK isn't important, so it should be stripped prior to insertion (so that we can just insert it at the end of the table).

**Comment by brookesy2** [comment]
Nope, no Pk. You can check formattedUL.CSV to see. I will double check to
be sure.

**Comment by brookesy2** [comment]
Just checked, no id column specified.

**Comment by Murodese** [comment]
Oh, that's probably why. My pkey sequence has a max value so far of like 11k, but the maximum PK in this table is 27k - probably because the import I did before included PKs (and hence didn't update the sequence). I'll update that and try again.

**Comment by Murodese** [comment]
Yeah, that was the problem. Works fine, now.

**Comment by brookesy2** [comment]
Yipee!!

**Comment by brookesy2** [comment]
Last commit, added some extra comments :)

**Comment by brookesy2** [comment]
Actually, one thing I don't do is fix the search names afterwards, as I'm not sure if we use it anywhere. Want me to add a loop to fix them up?

**Comment by Murodese** [comment]
Search names are (unsurprisingly) used for searches, so yeah, they should be fixed.

**Comment by brookesy2** [comment]
I more meant I don't think we use them out of the Pre table specifically, just releases.

I will fix it up now :)

**Comment by brookesy2** [comment]
Give that a try :)

**Comment by Murodese** [comment]
Might be better if you use the clean_release_name()? function from releases.py, just to maintain consistency.

**Comment by brookesy2** [comment]
Done. After this I need to submit the same fix for the prebot. As currently it cleans the name using the same method.

**Comment by Murodese** [comment]
Cool, thanks :) You might want to add this process as part of the install procedure as well?

**Comment by brookesy2** [comment]
That sounds like a good idea. 

Let me make a bugfix to prebot and the pre clean function. Then I will add this to install procedure.

Thanks for you patience as usual! :)
